src/mono/mono/metadata/marshal-ilgen.c

```diff
@@ -29,18 +29,21 @@ static void emit_string_free_icall (MonoMethodBuilder *mb, MonoMarshalConv conv)
 // TODO: Does this need to loose the mono_ prefix?
 static void mono_marshal_ilgen_legacy_init (void);
 
-static gboolean ilgen_cb_inited = FALSE;
-static MonoMarshalIlgenCallbacks ilgen_marshal_cb;
+static MonoMarshalIlgenCallbacks* ilgen_marshal_cb = NULL;
 
 void
 mono_install_marshal_callbacks_ilgen (MonoMarshalIlgenCallbacks *cb)
 {
-	g_assert (!ilgen_cb_inited);
```

Comment:
@jkotas Was reporting what looks like a race condition here: https://github.com/dotnet/runtime/issues/77090. I'm not sure exactly what interleaving can cause this though.

Reply:
@vargaz @naricc since e107bcbe1190b11a0ce91e06454efe079ec85796, `mono_marshal_ilgen_init` _may_ be called by embedders early, but it doesn't _have to be called_.  In the case where the driver doesn't call it early (for example the normal desktop mono driver doesn't call it), `get_marshal_cb` will initialize lazily.  The race is in the lazy case where one thread can see `ilgen_cb_inited == TRUE`, but the callbacks are all still null.

Reply:
Doesn't it get called during runtime startup ?

Reply:
> Doesn't it get called during runtime startup ?

I was surprised, but apparently it doesn't.  Going back to the very first PR it's always been lazy

Reply:
So the problem is that it's really doing 2 separate things:
1. it's picking if you want ilgen or noilgen marshaling (for example XI without interp wants noilgen, but if interp is on, they want ilgen)
2. it's installing the ilgen callbacks.

we can't install the callbacks until the embedder tells us if they want ilgen or noilgen.  Unfortunately there's only one function right now `mono_marshal_ilgen_init` which both signals "we want ilgen" and installs the ilgen callbacks.  To install the noilgen callbacks we wait until the first time that marshaling gets used.

So tl;dr we can't just call `mono_marshal_ilgen_init` from runtime initialization because that might not be what the user wants.

We could probably install the callbacks (either ilgen or noilgen) late during startup, and use `mono_marshal_ilgen_init` just to set a flag that we want ilgen. That's essentially what Nathan's componentization work will do.  but that's actually a riskier change than this.  And we might need to backport to 7 if this race condition shows up there, too.

Reply:
Well I don't have any objection to just using `mono_marshal_ilgen_init` to set a flag and doing all the callback initialization late during startup.  @naricc I think that's what you wanted to do for the component in the first place right?

Reply:
@lambdageek That is what I am doing for the componentization change.  

So is that what we should do here too, instead of the thing with atomics? `mono_marshal_ilgen_init` sets a flag; we check that flag during, say, `mini_init()`. At which point we do callback installation if it is set?

Reply:
I guess in the case that we are using NOILGEN and no one has called `mono_marshal_ilgen_init` we just want to go ahead and install the noilgen callbacks?

Reply:
> 

Yea that sounds right.

> So is that what we should do here too, instead of the thing with atomics? mono_marshal_ilgen_init sets a flag; we check that flag during, say, mini_init(). At which point we do callback installation if it is set?

Yea, basically at the point where in the componentized versino we'd call the component init function, here we'd just set up the callbacks.

I think writing down https://github.com/dotnet/runtime/pull/77383#discussion_r1003661360 helped me to think about what the real problem is.   It's not concurrency.  It's that we have a confusing API for choosing between ilgen/noilgen.  We should init the marshaling stack at startup - there's no need for it to be lazy.   You were right in your proposal about `mono_marshal_ilgen_init` (just use it to set a flag), but I didn't understand the problem properly.